### ISSUE TITLE
Release cadence for dd-lib-js-init Docker Image?

### ISSUE THREAD

**Issue by MattMencel** [DataDog/dd-trace-js#2677]
I see 3.10.0 was released here on GitHub last week. What is the process for the init image for this release being built and published to dockerhub, GCR, AWS Gallery?

I've begun using the Admission Controller Injection method and would like to upgrade if possible. I was seeing some higher memory usage in 3.9.1 that I think was at least partly addressed in the new image.

**Comment by rossmotley** [comment]
I concur - Is there anywhere which documents the release cadence of `dd-lib-js-init`? In my case, I am keen for the fix for  https://github.com/DataDog/dd-trace-js/issues/2593 which came in v3.9.2 on Dec 6, 2022.

I have followed the [library injection](https://docs.datadoghq.com/tracing/trace_collection/admission_controller/)" approach for instrumenting my apps, but if it takes a few weeks for it to be updated, it's probably best avoided.

**Comment by tlhunter** [comment]
I see that 3.10.0 image was released [four days ago](https://registry.hub.docker.com/r/datadog/dd-lib-js-init/tags), which is about three days after the library was released. I suspect it's automated somehow but I'm not sure of the exact mechanism. I'll ask around.

**Comment by tlhunter** [comment]
The list of images is available here:
https://console.cloud.google.com/gcr/images/datadoghq/GLOBAL/dd-lib-js-init

I looked into this a bit more. The process for releasing these images is automatic. Whenever we push up a new release it should automatically build and release the docker image. This is handled in [.github/workflows/release-latest.yml](https://github.com/DataDog/dd-trace-js/blob/7fb068aff5a877360e968523935811832ae7a75f/.github/workflows/release-latest.yml#L34).

The question then is two fold. First, why are there missing releases like 3.9.2 and 3.9.3, and the second, why is 3.11.1 not also tagged with latest?

While spelunking through the complex GitHub action output, looking back between the 3.9.2 (unlisted) and the 3.10.0 (listed) releases, I did notice that the build output differs slightly:

### v3.9.3

<img width="1144" alt="Screenshot 2023-01-19 at 13 05 52" src="https://user-images.githubusercontent.com/551402/213559694-44961026-f717-4394-b5ee-de35f2c49dd2.png">

### v3.10.0

<img width="1143" alt="Screenshot 2023-01-19 at 13 05 44" src="https://user-images.githubusercontent.com/551402/213559728-6a165198-ad84-4561-8800-a9a5132fbbe1.png">

The `docker buildx build` commands have different `--push` flags, with 3.9.3 being `k8s/auto-inst` and with 3.10.0 being `lib-injection`.

There's another difference in output:

### v3.9.3

```
#1 [internal] booting buildkit
#1 pulling image moby/buildkit:buildx-stable-1 0.1s done
#1 creating container buildx_buildkit_lib-injection0
#1 creating container buildx_buildkit_lib-injection0 0.5s done
#1 DONE 0.7s
2022/12/09 17:43:49 http2: server connection error from localhost: connection error: PROTOCOL_ERROR
...
#27 DONE 12.5s
```

### v3.10.0

```
#1 [internal] booting buildkit
#1 pulling image moby/buildkit:buildx-stable-1
#1 pulling image moby/buildkit:buildx-stable-1 0.8s done
#1 creating container buildx_buildkit_lib-injection0
#1 creating container buildx_buildkit_lib-injection0 0.7s done
#1 DONE 1.5s
...
#36 DONE 15.4s
```

The engineer who originally set this up is currently OOO so I'm still trying to track things down.

**Comment by tlhunter** [comment]
The step that I showed in screenshots is actually handled in this other file, [.github/actions/injection/action.yml](https://github.com/DataDog/dd-trace-js/blob/7fb068aff5a877360e968523935811832ae7a75f/.github/actions/injection/action.yml).

The step in the YAML file looks like this:

```sh
docker buildx build \
  --platform=linux/amd64,linux/arm/v7,linux/arm64/v8 \
  -t ghcr.io/datadog/dd-trace-js/dd-lib-js-init:${{ inputs.init-image-version }} \
  --push lib-injection
```

The interpreted command output for 3.9.3 looks like this:

```sh
docker buildx build \
  --platform=linux/amd64,linux/arm/v7,linux/arm64/v8 \
  -t ghcr.io/datadog/dd-trace-js/dd-lib-js-init:v3.9.3 \
  --push k8s/auto-inst
```

The interpreted command output for 3.10.0 looks like this:

```sh
docker buildx build \
  --platform=linux/amd64,linux/arm/v7,linux/arm64/v8 \
  -t ghcr.io/datadog/dd-trace-js/dd-lib-js-init:v3.10.0 \
  --push lib-injection
```

It looks like the push target was changed in this pull request: https://github.com/DataDog/dd-trace-js/pull/2544/files

**Comment by robertomonteromiguel** [comment]
Yes, a few month ago we changed the lib-injection image generation procedure and we unified the tests on system-tests.
The idea behind this was test all lib-injection images (for each language, java, python and js) in the same way.
We also try not to make radical changes to the tagging of the images, as they are published artifacts. Only the necessary changes to homogenize the generation and testing of images.
These images we are publishing to [ghcr.io](http://ghcr.io/).

I can see these last tags in ghcr:

- https://github.com/DataDog/dd-trace-js/pkgs/container/dd-trace-js%2Fdd-lib-js-init/57489240?tag=v3.9.2
- https://github.com/orgs/DataDog/packages/container/dd-trace-js%2Fdd-lib-js-init/58094664?tag=v3.9.3
- https://github.com/DataDog/dd-trace-js/pkgs/container/dd-trace-js%2Fdd-lib-js-init/61472338?tag=v3.10.0
- https://github.com/DataDog/dd-trace-js/pkgs/container/dd-trace-js%2Fdd-lib-js-init/62963607?tag=v3.11.0
- 
I don’t know how the releases are published to Docker Hub: https://hub.docker.com/r/datadog/dd-lib-js-init/tags
I don’t know how the images are deployed here: https://console.cloud.google.com/gcr/images/datadoghq/GLOBAL/dd-lib-js-init. 

I would like to know how is the process that publishes the images in other registry different from ghcr…

**Comment by robertomonteromiguel** [comment]
Could the problem have happened here?

https://gitlab.ddbuild.io/DataDog/apm-reliability/dd-trace-js/-/pipelines/11728769

**Comment by tlhunter** [comment]
We ended up discovering a race condition between the time it takes GitHub to build and push an image and the time it takes GitLab to pull and distribute the image. I even had to make a diagram of the problem to understand it so I figured why not upload it here:

![gl-gh-k8s-flow](https://user-images.githubusercontent.com/551402/218171752-36053ca6-613b-4fed-bb2a-82d2aa68c092.png)

The solution that we came up with is to retry to pull the image from GitLab with a delay between each retry. That was deployed as of #2785. It will retry up to five times with a delay in-between of 5 minutes. 25 minutes should provide ample time for the build and push process to complete. With the next releases containing this fix we should see that the releases are now automatic and dependable.
